### DISCUSSION TITLE
Problem with MONAIlabel logging.json permission

### DISCUSSION THREAD

**Discussion by OmarAshkar** [Project-MONAI/MONAILabel#1452] [Q&A]
Hello,

I am using the radiology app with 3d slicer, but users being switched to connect to a server, we get an error for logging.json file permission. we have to delete it manually. 

Is there is a way to fix that? Also generally is the current radiology app ready for multiple users to annotate the same time?

Your guys work is highly appreciated ^^ 

Thank you,
Omar

**Comment by SachidanandAlle** [comment]
logging.json?  can you please provide more details..
- from which path?
- client side or server side issue?
- users switch? didn't understand.. 
- may be some snapshot? exception trace? logs?

**Reply by OmarAshkar** [reply]
@SachidanandAlle 
- apps/radiology/logs/logging.json'
- server side (server not initiallizing)
- User A start the server. User B cannot start the server unless user A exit and the logging.json file is deleted.


```
2023-05-25 18:21:58,571 - ENV SETTINGS:: MONAI_LABEL_AUTO_UPDATE_SCORING = True
2023-05-25 18:21:58,571 - 
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/conda/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/opt/conda/lib/python3.8/site-packages/monailabel/main.py", line 342, in <module>
    Main().run()
  File "/opt/conda/lib/python3.8/site-packages/monailabel/main.py", line 133, in run
    self.action_start_server(args)
  File "/opt/conda/lib/python3.8/site-packages/monailabel/main.py", line 229, in action_start_server
    log_config = init_log_config(args.log_config, args.app, "app.log", args.verbose)
  File "/opt/conda/lib/python3.8/site-packages/monailabel/utils/others/generic.py", line 105, in init_log_config
    shutil.copy(default_config, log_config)
  File "/opt/conda/lib/python3.8/shutil.py", line 418, in copy
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/opt/conda/lib/python3.8/shutil.py", line 264, in copyfile
    with open(src, 'rb') as fsrc, open(dst, 'wb') as fdst:
PermissionError: [Errno 13] Permission denied: '/workspace/apps/radiology/logs/logging.json'
```
Hope this was clear. 

So my question is how to fix, and if users are allowed to let's say annotate the same image at the same time, how these conflicts will be resolved.

Thanks!

**Comment by SachidanandAlle** [comment]
Why there are multiple start?  I am trying to understand why both User A and User B have to start the server.
User A and User B both can share the same server (started once)

When you fetch Next Sample, user A will get a different image and user B will next one.. unless that is the last image remains.. (it's not guaranteed as it random selection is simply based on weighted probability)

However you can hook in your own strategy if you want to handle distributed user-case.  Client-ID is always passed in the request and you can maintain a session based on the same.  

In the examples, please remember as part of open source, we support only stateless instance and still try to support simple distribution across multiple users.  However you have an option to extend these Next Sample (strategy) into your own..

For example: https://github.com/Project-MONAI/MONAILabel/blob/main/monailabel/tasks/activelearning/random.py#L40
And you can extend to support a bit better based on the multiple clients or simple based on client-id (coming in request arg) instead of weighted probability based on last time access..

When you such strategy, then you can use the same in your App.
Example: https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/radiology/main.py#L225


In any case if you want to go with pre-distribution strategy where same app you want to run with different studies, then you can provide your log config while starting the server.. something like this..

`monailabel startserver --help`

`monilabel startserver -s s1 -a radiology -l user1.log.json`
`monilabel startserver -s s2 -a radiology -l user2_log.json`

May be easiest way to have your own log config is by copying the `/workspace/apps/radiology/logs/logging.json` and modify references of `app.log` into say `user1.log`


Note: this is a cool idea to distribute the work across multiple users but not efficient wrt GPU.. as each process (User) will load the same model multiple times

**Reply by OmarAshkar** [reply]
Thank you @SachidanandAlle. The thing is the server is not sustainable and users needs to disconnect and reconnect, so there is no single source for it. 

Additionally, I tried the `-l` argument but it didn't work unfortunately.

Thank you

**Comment by SachidanandAlle** [comment]
will check on this and if needed, we can have a fix for that..
to unblock yourself you can try different copies of application (radiology)..

**Comment by SachidanandAlle** [comment]
^^ @tangy5

**Comment by tangy5** [comment]
Recently, we have enabled an option to conduct multi-user authentication for different roles, say Reviewer, Admin, Annotator, User can work on same project and server. 

We haven't seen this logging permission issue, I will investigate it, thanks for posting this. 

i will try reproduce this error, if convenient, can you provide a simple example (steps) of how to get this error? Thanks so much!

**Reply by OmarAshkar** [reply]
Thank you @tangy5.  Just to clarify. Each user start the server himself and start annotating. So, each one write the file as owner and for some reason, the permission does not allow other group members to write on that.

**Reply by OmarAshkar** [reply]
Hi @tangy5. As this problem became very nagging to me. Is there is a flag to specify logging should be saved per single server start or automatically override the logging.json? Whenever a user starts a server he creates logging.json with permission -rw-r--r--. It doesn't matter if this permission has changed, no one can write on this file afterwards.

**Comment by SachidanandAlle** [comment]
@OmarAshkar can you try with this branch.. 
https://github.com/Project-MONAI/MONAILabel/tree/log_fix

if this solves the problem for your.. we can merge this into main branch and possibly release this fix in 0.7 release... (0.7 release is due by tomorrow)

**Comment by SachidanandAlle** [comment]
@OmarAshkar were you able to test/verify the patch?
https://github.com/Project-MONAI/MONAILabel/commit/2cf6bed937d73b92e18de185e2a7a5ef664b585a

**Comment by OmarAshkar** [comment]
@SachidanandAlle Yes I confirm that the file is not generated on the server directory anymore. 

Thanks a lot!
